lib/spack/spack/provider_index.py

```diff
@@ -2,54 +2,134 @@
 # Spack Project Developers. See the top-level COPYRIGHT file for details.
 #
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
+"""Classes and functions to manage providers of virtual dependencies"""
+import itertools
 
-"""
-The ``virtual`` module contains utility classes for virtual dependencies.
-"""
-
-from itertools import product as iproduct
-from six import iteritems
-from pprint import pformat
-
+import six
 import spack.error
 import spack.util.spack_json as sjson
 
 
-class ProviderIndex(object):
-    """This is a dict of dicts used for finding providers of particular
-       virtual dependencies. The dict of dicts looks like:
+def _cross_provider_maps(lmap, rmap):
+    """Return a dictionary that combines constarint requests from both input.
 
-       { vpkg name :
-           { full vpkg spec : set(packages providing spec) } }
+    Args:
+        lmap: main provider map
+        rmap: provider map with additional constraints
+    """
+    # TODO: this is pretty darned nasty, and inefficient, but there
+    # TODO: are not that many vdeps in most specs.
+    result = {}
+    for lspec, rspec in itertools.product(lmap, rmap):
+        try:
+            constrained = lspec.constrained(rspec)
+        except spack.error.UnsatisfiableSpecError:
+            continue
+
+        # lp and rp are left and right provider specs.
+        for lp_spec, rp_spec in itertools.product(lmap[lspec], rmap[rspec]):
+            if lp_spec.name == rp_spec.name:
+                try:
+                    const = lp_spec.constrained(rp_spec, deps=False)
+                    result.setdefault(constrained, set()).add(const)
+                except spack.error.UnsatisfiableSpecError:
+                    continue
+    return result
+
+
+class _IndexBase(object):
+    #: This is a dict of dicts used for finding providers of particular
+    #: virtual dependencies. The dict of dicts looks like:
+    #:
+    #:    { vpkg name :
+    #:        { full vpkg spec : set(packages providing spec) } }
+    #:
+    #: Callers can use this to first find which packages provide a vpkg,
+    #: then find a matching full spec.  e.g., in this scenario:
+    #:
+    #:    { 'mpi' :
+    #:        { mpi@:1.1 : set([mpich]),
+    #:          mpi@:2.3 : set([mpich2@1.9:]) } }
+    #:
+    #: Calling providers_for(spec) will find specs that provide a
+    #: matching implementation of MPI. Derived class need to construct
+    #: this attribute according to the semantics above.
+    providers = None
+
+    def providers_for(self, virtual_spec):
+        """Return a list of specs of all packages that provide virtual
+        packages with the supplied spec.
+
+        Args:
+            virtual_spec: virtual spec to be provided
+        """
+        result = set()
+        # Allow string names to be passed as input, as well as specs
+        if type(virtual_spec) == str:
```

Comment:
Is this preferred over `isinstance(virtual_spec, str)`?

Reply:
No, that's just what was there before. I'll change it to:
```python
if isinstance(virtual_spec, six.string_types):
```

Reply:
I'll have to remember `six.string_types`.